pgjdbc/src/main/java/org/postgresql/jdbc/PgPreparedStatement.java

```diff
@@ -877,33 +878,64 @@ private static BigDecimal castToBigDecimal(final Object in, final int scale) thr
   }
 
   private static boolean castToBoolean(final Object in) throws SQLException {
-    try {
-      if (in instanceof String) {
-        return ((String) in).equalsIgnoreCase("true") || ((String) in).equals("1")
-            || ((String) in).equalsIgnoreCase("t");
-      }
-      if (in instanceof BigDecimal) {
-        return ((BigDecimal) in).signum() != 0;
-      }
-      if (in instanceof Number) {
-        return ((Number) in).longValue() != 0L;
-      }
-      if (in instanceof java.util.Date) {
-        return ((java.util.Date) in).getTime() != 0L;
-      }
-      if (in instanceof Boolean) {
-        return (Boolean) in;
+    if (in instanceof Boolean) {
+      return (Boolean) in;
+    }
+    if (in instanceof String) {
+      // Leading or trailing whitespace is ignored, and case does not matter.
+      Boolean value = null;
+      final String strval = ((String) in).trim();
+      if ("1".equals(strval) || "true".equalsIgnoreCase(strval)
+          || "t".equalsIgnoreCase(strval) || "yes".equalsIgnoreCase(strval)
+          || "y".equalsIgnoreCase(strval) || "on".equalsIgnoreCase(strval)) {
+        value = Boolean.TRUE;
+      } else if ("0".equals(strval) || "false".equalsIgnoreCase(strval)
+          || "f".equalsIgnoreCase(strval) || "no".equalsIgnoreCase(strval)
+          || "n".equalsIgnoreCase(strval) || "off".equalsIgnoreCase(strval)) {
+        value = Boolean.FALSE;
+      } else if (value == null) {
+        throw cannotCastException("String", "boolean");
       }
-      if (in instanceof Clob) {
-        final String asString = asString((Clob) in);
-        return asString.equalsIgnoreCase("true") || asString.equals("1")
-            || asString.equalsIgnoreCase("t");
+      return value;
+    }
+    if (in instanceof Character) {
+      Boolean value = null;
+      final Character charval = (Character) in;
+      if ('1' == charval || 't' == charval || 'T' == charval
+          || 'y' == charval || 'Y' == charval) {
+        value = Boolean.TRUE;
+      } else if ('0' == charval || 'f' == charval || 'F' == charval
+          || 'n' == charval || 'N' == charval) {
+        value = Boolean.FALSE;
+      } else if (value == null) {
+        throw cannotCastException("Character", "boolean");
       }
-      if (in instanceof Character) {
-        return (Character) in == '1' || (Character) in == 't' || (Character) in == 'T';
+      return value;
+    }
+    if (in instanceof BigDecimal) {
+      return ((BigDecimal) in).signum() != 0;
+    }
+    if (in instanceof Number) {
+      return ((Number) in).longValue() != 0L;
+    }
+    if (in instanceof java.util.Date) {
+      return ((java.util.Date) in).getTime() != 0L;
+    }
+    if (in instanceof Clob) {
+      Boolean value = null;
+      final String strval = asString((Clob) in).trim();
+      if ("1".equals(strval) || "true".equalsIgnoreCase(strval)
```

Comment:
seems to me this is duplicate of the code above. I would think it should be possible to only write this once ?